crypto/ml_dsa/ml_dsa_key.c

```diff
@@ -21,31 +22,57 @@
  * @brief Create a new ML_DSA_KEY object
  *
  * @param libctx A OSSL_LIB_CTX object used for fetching algorithms.
+ * @param propq The property query used for fetching algorithms
  * @param alg The algorithm name associated with the key type
  * @returns The new ML_DSA_KEY object on success, or NULL on malloc failure
  */
-ML_DSA_KEY *ossl_ml_dsa_key_new(OSSL_LIB_CTX *libctx, const char *alg)
+ML_DSA_KEY *ossl_ml_dsa_key_new(OSSL_LIB_CTX *libctx, const char *propq,
+                                const char *alg)
 {
     ML_DSA_KEY *ret;
-    size_t poly_sz;
     const ML_DSA_PARAMS *params = ossl_ml_dsa_params_get(alg);
-    POLY *poly;
 
     if (params == NULL)
         return NULL;
 
-    poly_sz = sizeof(POLY) * (params->k * 3 + params->l);
-    ret = OPENSSL_zalloc(sizeof(*ret) + poly_sz);
+    ret = OPENSSL_zalloc(sizeof(*ret));
     if (ret != NULL) {
         ret->libctx = libctx;
         ret->params = params;
-        poly = (POLY *)((uint8_t *)ret + sizeof(*ret));
-        vector_init(&ret->t0, poly, params->k);
-        vector_init(&ret->t1, poly + params->k, params->k);
-        vector_init(&ret->s2, poly + 2 * params->k, params->k);
-        vector_init(&ret->s1, poly + 3 * params->k, params->l);
+        ret->shake128_md = EVP_MD_fetch(libctx, "SHAKE-128", propq);
+        ret->shake256_md = EVP_MD_fetch(libctx, "SHAKE-256", propq);
+        if (ret->shake128_md == NULL || ret->shake256_md == NULL)
+            goto err;
     }
     return ret;
+err:
+    ossl_ml_dsa_key_free(ret);
+    OPENSSL_free(ret);
```

Comment:
This is a double-free, the key is (appropriately) already freed in `ossl_ml_dsa_key_free`.

Reply:
oops,